homeassistant/config.py

```diff
@@ -578,26 +578,57 @@ def find_annotation_rec(
     return find_annotation_rec(config, list(path), None)
 
 
-def stringify_invalid(ex: vol.Invalid) -> str:
+def stringify_invalid(
+    ex: vol.Invalid,
+    domain: str,
+    config: dict,
+    link: str | None,
+    max_sub_error_length: int,
+) -> str:
     """Stringify voluptuous.Invalid.
 
     This is an alternative to the custom __str__ implemented in
-    voluptuous.error.Invalid. The main modification is to format
-    the path delimited by -> instead of @data[].
+    voluptuous.error.Invalid. The modifications are:
+    - Format the path delimited by -> instead of @data[]
+    - Prefix with domain, file and line of the error
+    - Suffix with a link to the documentation
+    - Give a more user friendly output for unknown options
     """
+    message_prefix = f"Invalid config for [{domain}]"
+    if domain != CONF_CORE and link:
+        message_suffix = f". Please check the docs at {link}"
+    else:
+        message_suffix = ""
+    if annotation := find_annotation(config, ex.path):
+        message_prefix += f" at {annotation[0]}, line {annotation[1]}"
     path = "->".join(str(m) for m in ex.path)
+    if ex.error_message == "extra keys not allowed":
+        return (
+            f"{message_prefix}: '{ex.path[-1]}' is an invalid option for [{domain}], "
+            f"check: {path}{message_suffix}"
+        )
     # This function is an alternative to the stringification done by
     # vol.Invalid.__str__, so we need to call Exception.__str__ here
     # instead of str(ex)
     output = Exception.__str__(ex)
     if error_type := ex.error_type:
         output += " for " + error_type
-    return f"{output} '{path}'"
+    offending_item_summary = repr(_get_by_path(config, ex.path))
+    if len(offending_item_summary) > max_sub_error_length:
+        offending_item_summary = (
+            f"{offending_item_summary[: max_sub_error_length - 3]}..."
+        )
+    return (
+        f"{message_prefix}: {output} '{path}', got {offending_item_summary}"
+        f"{message_suffix}."
+    )
 
 
 def humanize_error(
-    data: Any,
     validation_error: vol.Invalid,
+    domain: str,
+    config: dict,
+    link: str | None,
     max_sub_error_length: int = MAX_VALIDATION_ERROR_ITEM_LENGTH,
 ) -> str:
     """Provide a more helpful + complete validation error message.
```

Comment:
It would be nice if this improved helper would become part of the `voluptous` library.

Reply:
I think it's too targeted for Home Assistant.